src/lib.rs

```diff
@@ -116,40 +117,55 @@ pub struct Rules {
 }
 
 impl Rules {
-    /// Split a text into segments.
-    pub fn split<'a>(&self, text: &'a str) -> Vec<&'a str> {
+    /// Obtain the ranges for text segments. Guaranteed to be at character bounds.
+    pub fn split_ranges(&self, text: &str) -> Vec<Range<usize>> {
         let mut segments = Vec::new();
 
-        let mut mask: Vec<Option<bool>> = vec![None; text.len()];
+        // TODO use a proper tri-state enum here
+        let mut masked_bytes: Vec<Option<bool>> = vec![None; text.len()];
 
-        for rule in &self.rules {
-            for index in rule.match_indices(text) {
-                if index >= text.len() {
-                    continue;
+        'outer: for rule in &self.rules {
+            for byte_index in rule.match_indices(text) {
+
+                if byte_index >= text.len() {
+                    continue 'outer;
                 }
 
-                if mask[index].is_none() {
-                    mask[index] = Some(rule.do_break());
+                if masked_bytes[byte_index].is_none() {
+                    masked_bytes[byte_index] = Some(rule.do_break());
                 }
             }
         }
 
-        let mut prev_index = 0;
+        let mut prev_byte_pos = 0;
 
-        for (i, mask_val) in (0..text.len()).zip(mask) {
-            if let Some(true) = mask_val {
-                segments.push(&text[prev_index..i]);
-                prev_index = i;
+        // Iterate over characters, we don't want no half characters in the output ranges
+        for (byte_pos, _c) in text.char_indices() {
+            if let Some(Some(true)) = masked_bytes.get(byte_pos) {
```

Comment:
I'd prefer indexing here instead of `.get` since this is a bug if it fails. Looks great otherwise, thanks!